File: slice_cli/context.py
from __future__ import annotations

import os
import subprocess
from pathlib import Path


class Ctx:
    """Lazily-resolved repo paths. Create once, pass everywhere."""

    def __init__(self, repo: str | None = None, slices_dir: str | None = None):
        self._repo = Path(repo).resolve() if repo else None
        self._slices = Path(slices_dir).resolve() if slices_dir else None

    @property
    def repo_root(self) -> Path:
        if self._repo is None:
            env = os.environ.get("SLICES_REPO_ROOT")
            if env:
                self._repo = Path(env).resolve()
            else:
                current = Path.cwd()
                for parent in [current, *current.parents]:
                    if (parent / ".git").exists():
                        self._repo = parent
                        break
                else:
                    raise RuntimeError(
                        "Not inside a git repository. Set SLICES_REPO_ROOT or use --repo."
                    )
        return self._repo

    @property
    def slices_dir(self) -> Path:
        if self._slices is None:
            env = os.environ.get("SLICES_DIR")
            self._slices = Path(env).resolve() if env else self.repo_root / "slices"
        return self._slices

    @property
    def index_path(self) -> Path:
        return self.slices_dir / "INDEX.md"

    @property
    def docs_manifest_path(self) -> Path:
        return self.slices_dir / "DOCS.yaml"

    def git(self, *args: str, check: bool = True) -> subprocess.CompletedProcess[str]:
        return subprocess.run(
            ["git", *args],
            cwd=self.repo_root,
            capture_output=True,
            text=True,
            check=check,
        )

    def head_sha(self) -> str:
        try:
            return self.git("rev-parse", "HEAD").stdout.strip()
        except (OSError, subprocess.CalledProcessError):
            return "unknown"


    def rel(self, path: Path) -> str:
        try:
            return str(path.relative_to(self.repo_root))
        except ValueError:
            return str(path)

File: rust/slice-rs/src/context.rs
use std::env;
use std::path::{Path, PathBuf};

use crate::{Error, Result};

#[derive(Debug, Clone)]
pub struct Context {
    repo_root: PathBuf,
    git_root: PathBuf,
    slices_dir: PathBuf,
}

impl Context {
    #[cfg(test)]
    #[must_use]
    pub(crate) fn from_parts_for_test(
        repo_root: PathBuf,
        git_root: PathBuf,
        slices_dir: PathBuf,
    ) -> Self {
        Self {
            repo_root,
            git_root,
            slices_dir,
        }
    }

    pub fn new(repo: Option<PathBuf>, slices_dir: Option<PathBuf>) -> Result<Self> {
        let repo_root = match repo {
            Some(path) => path
                .canonicalize()
                .map_err(|source| Error::Read { path, source })?,
            None => discover_repo_root()?,
        };
        let slices_dir = match slices_dir {
            Some(path) => path
                .canonicalize()
                .unwrap_or_else(|_| absolutize(&path).unwrap_or(path)),
            None => repo_root.join("slices"),
        };
        let git_root = discover_ancestor_git_root(&repo_root).unwrap_or_else(|| repo_root.clone());
        Ok(Self {
            repo_root,
            git_root,
            slices_dir,
        })
    }

    #[must_use]
    pub fn repo_root(&self) -> &Path {
        &self.repo_root
    }

    #[must_use]
    pub fn git_relative_path(&self, repo_relative: &str) -> String {
        self.repo_root.strip_prefix(&self.git_root).map_or_else(
            |_| repo_relative.to_owned(),
            |prefix| prefix.join(repo_relative).to_string_lossy().into_owned(),
        )
    }

    #[must_use]
    pub fn slices_dir(&self) -> &Path {
        &self.slices_dir
    }

    #[must_use]
    pub fn docs_manifest_path(&self) -> PathBuf {
        self.slices_dir.join("DOCS.yaml")
    }

    #[must_use]
    pub fn index_path(&self) -> PathBuf {
        self.slices_dir.join("INDEX.md")
    }

    #[must_use]
    pub fn head_sha(&self) -> String {
        let Ok(output) = std::process::Command::new("git")
            .args(["rev-parse", "HEAD"])
            .current_dir(&self.repo_root)
            .output()
        else {
            return "unknown".to_owned();
        };
        if output.status.success() {
            String::from_utf8_lossy(&output.stdout).trim().to_owned()
        } else {
            "unknown".to_owned()
        }
    }

    #[must_use]
    pub fn rel(&self, path: &Path) -> String {
        path.strip_prefix(&self.repo_root).map_or_else(
            |_| path.to_string_lossy().into_owned(),
            |p| p.to_string_lossy().into_owned(),
        )
    }
}

fn discover_repo_root() -> Result<PathBuf> {
    if let Ok(env_root) = env::var("SLICE_REPO") {
        return PathBuf::from(env_root)
            .canonicalize()
            .map_err(|source| Error::Read {
                path: PathBuf::from("SLICE_REPO"),
                source,
            });
    }
    if let Ok(env_root) = env::var("SLICES_REPO_ROOT") {
        return PathBuf::from(env_root)
            .canonicalize()
            .map_err(|source| Error::Read {
                path: PathBuf::from("SLICES_REPO_ROOT"),
                source,
            });
    }

    let mut current = env::current_dir()?;
    loop {
        if current.join(".git").exists() {
            return Ok(current);
        }
        if !current.pop() {
            return Err(Error::NotInRepo);
        }
    }
}

fn discover_ancestor_git_root(start: &Path) -> Option<PathBuf> {
    let mut current = start.to_path_buf();
    loop {
        if current.join(".git").exists() {
            return Some(current);
        }
        if !current.pop() {
            return None;
        }
    }
}

fn absolutize(path: &Path) -> std::io::Result<PathBuf> {
    if path.is_absolute() {
        Ok(path.to_path_buf())
    } else {
        Ok(env::current_dir()?.join(path))
    }
}

#[cfg(test)]
mod tests {
    use super::Context;

    #[test]
    fn paths_are_derived_from_context_roots() {
        let repo = std::path::PathBuf::from("/tmp/work/repo");
        let git = std::path::PathBuf::from("/tmp/work");
        let ctx = Context::from_parts_for_test(repo.clone(), git, repo.join("custom-slices"));

        assert_eq!(
            ctx.docs_manifest_path(),
            repo.join("custom-slices/DOCS.yaml")
        );
        assert_eq!(ctx.rel(&repo.join("src/lib.rs")), "src/lib.rs");
        assert_eq!(ctx.git_relative_path("src/lib.rs"), "repo/src/lib.rs");
    }

    #[test]
    fn head_sha_is_unknown_outside_git_repo() {
        let temp = tempfile::tempdir().unwrap();
        let ctx = Context::from_parts_for_test(
            temp.path().to_path_buf(),
            temp.path().to_path_buf(),
            temp.path().join("slices"),
        );
        assert_eq!(ctx.head_sha(), "unknown");
    }
}
